spec/asyncapi.md

```diff
@@ -694,37 +696,97 @@ bindings:
 
 
 
+#### <a name="operationsObject"></a>Operations Object
 
+Holds a dictionary with all the [operations](#operationObject) this application MUST implement.
 
-#### <a name="operationObject"></a>Operation Object
+Note: If you're looking for a place to define operations that MAY or MAY NOT be implemented by the application, consider defining them in [`components/operations`](#componentsOperations).
+
+##### Patterned Fields
+
+Field Pattern | Type | Description
+---|:---:|---
+<a name="operationsObjectOperation"></a>{operationId} | [Operation Object](#channelItemObject) \| [Reference Object](#referenceObject) | The operation this application MUST implement. The field name (`operationId`) MUST be a string used to identify the operation in the document where it is defined, and its value is **case-sensitive**. Tools and libraries MAY use the `operationId` to uniquely identify an operation, therefore, it is RECOMMENDED to follow common programming naming conventions.
 
-Describes a publish or a subscribe operation. This provides a place to document how and why messages are sent and received.
+##### Operations Object Example
+
+```json
+{
+  "onUserSignUp": {
+    "summary": "Action to sign a user up.",
+    "description": "A longer description",
+    "channel": {
```

Comment:
I'm thinking that it might make sense to convert this into an array of channels (of references to channel) instead.
The use case that came to my mind was the **fanout** pattern,  sending the same message to different channels.

Reply:
AFAIK, the fanout pattern is usually implemented by the broker, not the application. I mean, for instance, RabbitMQ has the fanout exchange that lets you connect various queues to a specific routingKey. Kafka, NATS, and others have consumer groups, with which you can implement the fanout pattern. The application sends a message to a particular channel and then it gets distributed to many other queues or consumers. Am I missing something?

Now, I see the value it can have by allowing us to group multiple operations into one. Batch operations if you want. We just need to be very clear that if a `send` operation is performed against 1 out of 3 channels and it fails, it shouldn't stop the code from sending the other 2. Aren't we getting into the land of workflow definition already? 🤔

Reply:
> AFAIK, the fanout pattern is usually implemented by the broker, not the application.

This statement doesn't reflect the reality out there. The fanout could be perfectly implemented by the application; writing the same message to different Kafka topics is a completely valid use case. For logistics, you might want to set up different retention policies, permissions, etc to each topic.

Also, we can still use an AsyncAPI document to describe a message broker (wich turns to be the application), where this pattern is way more common.
At the end, a client should know where to connect to and to which channels a particular message will be sent so it can subscribe to one of those (depending on each scenario).

> Aren't we getting into the land of workflow definition already?

TBH I don't think we are that far. This is still referring to an application definition.

Reply:
Yeah, that's my point. I wouldn't say it's to support fanout but to batch operations. That's a pretty valid and common use case.

@derberg @dalelane @char0n @magicmatatjahu what do you folks think?

Reply:
I don't get why it's not to support the [fanout pattern](https://en.wikipedia.org/wiki/Fan-out_(software)). Messages are sent in parallel to different channels.

From Wikipedia:

>In message-oriented middleware solutions, fan-out is a messaging pattern used to model an information exchange that implies the delivery (or spreading) of a message to one or multiple destinations possibly in parallel, and not halting the process that executes the messaging to wait for any response to that message.

Reply:
Sorry, my point was that you can still do fanout without making it an array. In any case, check that definition mate, it clearly says "to model an **information exchange**" 😄 Exchanges are always part of the broker. If your application is sending the same message three times to three different destinations, it's not fanout, it's just a single publish three times. The advantage of fanout is that you only publish once and the message is spread to multiple destinations. I only have to succeed to send the message once and I only need to know about one destination. If your application is sending a message 3 times:

1. It needs to know the 3 destinations. Not a constraint for fanout but definitely misses one of the strongest points of fanout.
2. You need to make sure your application doesn't halt in the middle of the process, which may not be always possible, making the "fanout" unreliable.

That said, I don't care. We seem to understand each other, we just don't agree on what fanout means. Again, I think the proposal still has a point because it lets you batch/group operations. Whether we call it "fanout" or "batch operations" here, is not relevant for the discussion IMHO.

Reply:
Indeed we ended up talking about the same thing, each one believing strongly in its own point 😆. 
As you shared, let's don't care about how to call it but focus then on this possible use case that we should consider. 

Let's see what others say.

Reply:
it was indeed an interesting conversation,

I learned a lot.

Reply:
And what do you think about making it an array?

Reply:
> For example, a user sign up action on a server might result in the server sending the same message to several queues on an EDA. Let's say those queues are owned by different departments, teams, or whatever.,

I can't imagine such a situation, really. Knowing how things work, I bet that devs asked to make sure they send the same message to another dedicated channel for some team, will just tell people to subscribe to the existing channel 🤷🏼 

delivering one message to 2 channels is basically duplicating messages, right? cause they are different as they will have to have different tracing id?

Reply:
The only use case that comes to my mind right now is auditing. Some messages will be sent to their corresponding channel + the auditing channel. But I've never seen this done in the application code. It's always configured as a rule in the broker so you don't forget to audit anything.

Reply:
The use case I had in my mind is illustrated in the following pic grabbed from Salesforce blog. In particular, I'm talking about **Point-to-Point** architecture (even though the name is not correct IMHO) and not entering into which one is more scalable or less because it's not the main topic being discussed here.

![](https://d259t2jj6zp7qm.cloudfront.net/images/c_scale%2Cw_700-RealTimeIntegration_1_dmdeil.png)

Reply:
but point to point is actually in contrast to EDA, right?

Reply:
> but point to point is actually in contrast to EDA, right?

The name on the picture is not correct as I mentioned in my previous comment. The right name I'm looking for is **EventDriven Integration**, and this article from Solace explains what it is: https://solace.com/blog/event-driven-architecture-difference-event-driven-integration

Reply:
@smoya It looks like this requires its own discussion. Would you mind creating a follow-up issue? It would be great to find actual cases where this is required. Salesforce is great but I don't think it's a common use case.